File: examples/java-repo/src/test/java/com/example/ecommerce/application/order/PlaceOrderHandlerTest.java
package com.example.ecommerce.application.order;

import static org.junit.Assert.*;

import com.example.ecommerce.domain.customer.Customer;
import com.example.ecommerce.domain.customer.CustomerRepository;
import com.example.ecommerce.domain.inventory.InventoryRepository;
import com.example.ecommerce.domain.inventory.Stock;
import com.example.ecommerce.domain.order.Money;
import com.example.ecommerce.domain.order.Order;
import com.example.ecommerce.domain.order.OrderRepository;
import com.example.ecommerce.domain.product.Product;
import com.example.ecommerce.domain.product.ProductRepository;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicLong;
import org.junit.Before;
import org.junit.Test;

public class PlaceOrderHandlerTest {

    private PlaceOrderHandler handler;
    private Map<Long, Customer> customers;
    private Map<Long, Product> products;
    private Map<Long, Stock> stocks;
    private Map<Long, Order> orders;
    private AtomicLong orderIdSeq;

    @Before
    public void setUp() {
        customers = new HashMap<>();
        products = new HashMap<>();
        stocks = new HashMap<>();
        orders = new HashMap<>();
        orderIdSeq = new AtomicLong(1);

        customers.put(1L, new Customer(1, "Alice", "alice@example.com"));
        products.put(1L, new Product(1, "Laptop", "Gaming laptop", Money.of(999.99, "USD")));
        products.put(2L, new Product(2, "Mouse", "Wireless mouse", Money.of(29.99, "USD")));
        stocks.put(1L, new Stock(1, 10));
        stocks.put(2L, new Stock(2, 50));

        OrderRepository orderRepo = new OrderRepository() {
            @Override
            public void save(Order order) {
                orders.put(order.getId(), order);
            }

            @Override
            public Optional<Order> findById(long id) {
                return Optional.ofNullable(orders.get(id));
            }

            @Override
            public long nextId() {
                return orderIdSeq.getAndIncrement();
            }
        };

        CustomerRepository customerRepo = new CustomerRepository() {
            @Override
            public void save(Customer c) {
                customers.put(c.getId(), c);
            }

            @Override
            public Optional<Customer> findById(long id) {
                return Optional.ofNullable(customers.get(id));
            }
        };

        ProductRepository productRepo = id -> Optional.ofNullable(products.get(id));

        InventoryRepository inventoryRepo = new InventoryRepository() {
            @Override
            public void save(Stock s) {
                stocks.put(s.getProductId(), s);
            }

            @Override
            public Optional<Stock> findByProductId(long productId) {
                return Optional.ofNullable(stocks.get(productId));
            }
        };

        handler = new PlaceOrderHandler(orderRepo, customerRepo, productRepo, inventoryRepo);
    }

    @Test
    public void shouldPlaceOrderSuccessfully() {
        List<PlaceOrderHandler.OrderItemRequest> items = List.of(
            new PlaceOrderHandler.OrderItemRequest(1, 2),
            new PlaceOrderHandler.OrderItemRequest(2, 3));

        OrderDto result = handler.execute(1, items);

        assertNotNull(result);
        assertEquals(1, result.id);
        assertEquals(1, result.customerId);
        assertEquals("CONFIRMED", result.status);
        assertEquals(2, result.lines.size());
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldRejectUnknownCustomer() {
        handler.execute(999, List.of(new PlaceOrderHandler.OrderItemRequest(1, 1)));
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldRejectEmptyItems() {
        handler.execute(1, List.of());
    }

    @Test(expected = IllegalStateException.class)
    public void shouldRejectInsufficientStock() {
        handler.execute(1, List.of(new PlaceOrderHandler.OrderItemRequest(1, 100)));
    }
}

File: examples/go-repo/internal/application/order/place_order_handler_test.go
package order_test

import (
	"testing"

	apporder "github.com/example/go-repo/internal/application/order"
	"github.com/example/go-repo/internal/domain/customer"
	"github.com/example/go-repo/internal/domain/inventory"
	domainorder "github.com/example/go-repo/internal/domain/order"
	"github.com/example/go-repo/internal/domain/product"
)

type inMemoryOrderRepo struct {
	orders  map[int]domainorder.Order
	counter int
}

func (r *inMemoryOrderRepo) Save(o domainorder.Order) error {
	r.orders[o.ID()] = o
	return nil
}
func (r *inMemoryOrderRepo) FindByID(id int) (*domainorder.Order, error) {
	o, ok := r.orders[id]
	if !ok {
		return nil, nil
	}
	return &o, nil
}
func (r *inMemoryOrderRepo) NextID() (int, error) {
	r.counter++
	return r.counter, nil
}

type inMemoryCustomerRepo struct {
	customers map[int]customer.Customer
}

func (r *inMemoryCustomerRepo) Save(c customer.Customer) error {
	r.customers[c.ID()] = c
	return nil
}
func (r *inMemoryCustomerRepo) FindByID(id int) (*customer.Customer, error) {
	c, ok := r.customers[id]
	if !ok {
		return nil, nil
	}
	return &c, nil
}

type inMemoryProductRepo struct {
	products map[int]product.Product
}

func (r *inMemoryProductRepo) FindByID(id int) (*product.Product, error) {
	p, ok := r.products[id]
	if !ok {
		return nil, nil
	}
	return &p, nil
}
func (r *inMemoryProductRepo) Add(p product.Product) error {
	r.products[p.ID()] = p
	return nil
}

type inMemoryInventoryRepo struct {
	stocks map[int]inventory.Stock
}

func (r *inMemoryInventoryRepo) Save(s inventory.Stock) error {
	r.stocks[s.ProductID()] = s
	return nil
}
func (r *inMemoryInventoryRepo) FindByProductID(id int) (*inventory.Stock, error) {
	s, ok := r.stocks[id]
	if !ok {
		return nil, nil
	}
	return &s, nil
}

func setupHandler() *apporder.PlaceOrderHandler {
	orders := &inMemoryOrderRepo{orders: make(map[int]domainorder.Order)}
	customers := &inMemoryCustomerRepo{customers: make(map[int]customer.Customer)}
	products := &inMemoryProductRepo{products: make(map[int]product.Product)}
	inv := &inMemoryInventoryRepo{stocks: make(map[int]inventory.Stock)}

	c, _ := customer.NewCustomer(1, "Alice", "alice@example.com")
	customers.Save(c)

	laptopPrice, _ := domainorder.NewMoney(999.99, "USD")
	mousePrice, _ := domainorder.NewMoney(29.99, "USD")
	laptop, _ := product.NewProduct(1, "Laptop", "Gaming laptop", laptopPrice)
	mouse, _ := product.NewProduct(2, "Mouse", "Wireless mouse", mousePrice)
	products.Add(laptop)
	products.Add(mouse)

	s1, _ := inventory.NewStock(1, 10)
	s2, _ := inventory.NewStock(2, 50)
	inv.Save(s1)
	inv.Save(s2)

	return apporder.NewPlaceOrderHandler(orders, customers, products, inv)
}

func TestPlaceOrderSuccessfully(t *testing.T) {
	handler := setupHandler()
	items := []apporder.OrderItem{
		{ProductID: 1, Quantity: 2},
		{ProductID: 2, Quantity: 3},
	}
	result, err := handler.Execute(1, items)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result == nil {
		t.Fatal("expected non-nil result")
	}
	if result.CustomerID() != 1 {
		t.Errorf("expected customer ID 1, got %d", result.CustomerID())
	}
	if result.Status() != domainorder.StatusConfirmed {
		t.Errorf("expected confirmed, got %s", result.Status())
	}
	if len(result.Lines()) != 2 {
		t.Errorf("expected 2 lines, got %d", len(result.Lines()))
	}
}

func TestRejectUnknownCustomer(t *testing.T) {
	handler := setupHandler()
	_, err := handler.Execute(999, []apporder.OrderItem{{ProductID: 1, Quantity: 1}})
	if err == nil {
		t.Fatal("expected error for unknown customer")
	}
}

func TestRejectEmptyItems(t *testing.T) {
	handler := setupHandler()
	_, err := handler.Execute(1, []apporder.OrderItem{})
	if err == nil {
		t.Fatal("expected error for empty items")
	}
}

func TestRejectInsufficientStock(t *testing.T) {
	handler := setupHandler()
	_, err := handler.Execute(1, []apporder.OrderItem{{ProductID: 1, Quantity: 100}})
	if err == nil {
		t.Fatal("expected error for insufficient stock")
	}
}
